### PR TITLE
Map 1st floor

### PR DESCRIPTION
**Pull request by gparant**
# Front 
  - Add to the first floor map of TheCodingMachien Paris.
  - Start to create strategy, multi map and switch the players between map.
  
 # Back
 - Update strategy to create and share data by room of the map.
 - Create event to switch user sokect of Map in front and room in the back.

### CODE REVIEW

**Comment by moufmouf** [comment]
Rah, c'est trop fat! :D
Je n'ai pas le temps de creuser le code en détail, mais je viens de tester la démo et ça déchire :D 
:+1:

File: back/src/Controller/MapController.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        this.getMpas();
+        this.assetMaps();
+    }
+
+    assetMaps() {
+        this.App.use('/map/files', express.static('src/Assets/Maps'));
+    }
+
+    //permit to login on application. Return token to connect on Websocket IO.
+    getMpas() {
```
Comment (moufmouf) [inline back/src/Controller/MapController.ts:20]: Small type here:

```suggestion
    getMaps() {
```

File: back/src/Controller/MapController.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        this.App = App;
+        this.getMpas();
+        this.assetMaps();
+    }
+
+    assetMaps() {
+        this.App.use('/map/files', express.static('src/Assets/Maps'));
+    }
+
+    //permit to login on application. Return token to connect on Websocket IO.
```
Comment (moufmouf) [inline back/src/Controller/MapController.ts:19]:
```suggestion
    // Returns a map mapping map name to file name of the map
```

File: back/src/Middleware/AuthenticateMiddleware.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import {Application, Request, Response} from "express";
+import {BAD_REQUEST} from "http-status-codes";
+import Jwt, {JsonWebTokenError} from "jsonwebtoken";
+import {SECRET_KEY} from "../Enum/EnvironmentVariable";
+
+export class AuthenticateMiddleware{
```
Comment (moufmouf) [inline back/src/Middleware/AuthenticateMiddleware.ts:6]: This class is not used yet, right? Shall we keep it?

File: front/src/Phaser/Game/GameScene.ts [hunk 4]
```diff
@@ -95,10 +103,16 @@ export class GameScene extends Phaser.Scene implements GameSceneInterface{
         //add layer on map
         this.Layers = new Array<Phaser.Tilemaps.StaticTilemapLayer>();
         let depth = -2;
-        this.map.layers.forEach((layer) => {
+        this.map.layers.forEach((layer : ITiledMapLayer) => {
             if (layer.type === 'tilelayer') {
                 this.addLayer(this.Map.createStaticLayer(layer.name, this.Terrains, 0, 0).setDepth(depth));
             }
+            if (layer.type === 'tilelayer' && layer.name === "exit") {
+                this.loadNextGame(layer);
```
Comment (moufmouf) [inline front/src/Phaser/Game/GameScene.ts:111]: :+1: yay!

Comment (moufmouf) [reply front/src/Phaser/Game/GameScene.ts:111]: We should document that in the README (how to create a map with a start and an exit) => this is needed if we want some help for the maps

File: front/src/Phaser/Game/GameScene.ts [hunk 5]
```diff
@@ -120,14 +134,16 @@ export class GameScene extends Phaser.Scene implements GameSceneInterface{
...
 
-
         //initialise camera
         this.initCamera();
 
 
         // Let's generate the circle for the group delimiter
 
         this.circleTexture = this.textures.createCanvas('circleSprite', 96, 96);
+        if(!this.circleTexture || this.circleTexture.context){
```
Comment (moufmouf) [inline front/src/Phaser/Game/GameScene.ts:144]: This is breaking the display of the circle, because `this.circleTexture.context` should always evaluate to true.
Why did you need this? Is something breaking?

File: front/src/Phaser/Game/GameScene.ts [hunk 6]
```diff
@@ -138,6 +154,62 @@ export class GameScene extends Phaser.Scene implements GameSceneInterface{
...
+        if(!game){
+            return;
+        }
+
+        let tiles : any = layer.data;
+        tiles.forEach((objectKey : number, key: number) => {
+            if(objectKey === 0){
+                return;
+            }
+            let y = (key / 45);
```
Comment (moufmouf) [inline front/src/Phaser/Game/GameScene.ts:182]: This "45" and the "46" below seems tied to the map. Where do they come from?

**Comment by gulien** [comment]
Stylé 💯  ! Il manque l'amphi par contre et l'escalier est sensé être à l'entrée #detailsMatters

File: front/src/Phaser/Game/GameScene.ts (hunk 5 above)
Comment (gparant) [reply front/src/Phaser/Game/GameScene.ts:144]: Yes the changes have break circle feature. I will check and try to fix it.

File: README.md [hunk 7]
```diff
@@ -45,6 +45,12 @@ A few things to notice:
 
 ![](doc/images/tiled_screenshot_1.png)
 
+If you have exit scene
+- You must create layer "exit". The layer have cases where the gamer can switch to the next scene.
+- In layer properties, you must add "exitSceneUrl" property. It represents a url map of the next scene. Be careful, if you want that the next map will be correctly loaded, you must check that the map files are in folder  `back/src/Assets/Maps/<your map folder>`. The files will be accessible by url `<HOST>/map/files/<your map folder>/...`.
```
Comment (moufmouf) [inline README.md:50]: Wait... if a layer must be called "exit" and must have a "exitSceneUrl", that means one scene can only have one exit, right? Shouldn't we change this?

File: front/src/Phaser/Login/LogincScene.ts [hunk 8]
```diff
@@ -132,4 +152,16 @@ export class LogincScene extends Phaser.Scene implements GameSceneInterface {
     shareUserPosition(UsersPosition: import("../../Connexion").MessageUserPositionInterface[]): void {
         throw new Error("Method not implemented.");
     }
+
+    deleteGroup(groupId: string): void {
+        throw new Error("Method not implemented.");
+    }
+
+    shareGroupPosition(groupPositionMessage: GroupCreatedUpdatedMessageInterface): void {
+        throw new Error("Method not implemented.");
```
Comment (moufmouf) [inline front/src/Phaser/Login/LogincScene.ts:161]: Why do we need this? This is because of an interface, right? Is it logical that LogincScene implements this interface?

File: back/src/Controller/MapController.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        this.assetMaps();
+    }
+
+    assetMaps() {
+        this.App.use('/map/files', express.static('src/Assets/Maps'));
+    }
+
+    // Returns a map mapping map name to file name of the map
+    getMaps() {
+        this.App.get("/maps", (req: Request, res: Response) => {
```
Comment (moufmouf) [inline back/src/Controller/MapController.ts:21]: We should probably rename this endpoint to something like "/start-map"

File: back/src/Model/Websocket/ExtRooms.ts [hunk 10]
```diff
@@ -40,7 +41,11 @@ let RefreshUserPositionFunction = function(rooms : ExtRooms, Io: socketIO.Server
 
         // update position in the worl
         let messageUserPosition = new MessageUserPosition(data);
-        World.updatePosition(messageUserPosition);
+        let world = Worlds.get(messageUserPosition.roomId);
```
Comment (moufmouf) [inline back/src/Model/Websocket/ExtRooms.ts:44]: We could store the roomId directly in the socket of the user. This way, there is no need to send the roomId on the wire on every move.

File: README.md (hunk 7 above)
Comment (gparant) [reply README.md]: No, you can create two layers with the same name "exit" (the id layer will be different) and add multi tile to exit. 
One tile cannot behave the param "exitSceneUrl" because in this case, yes, one tile = one exit and in the TileSet we should have each tileset to each exit scene 👎.
To have multi exit on the map, we must use the layers with multi tileset what represents exit case for the gamer.
So one layer = one scene exit, but we could have infinite "exit" layers on the map.

File: front/src/Phaser/Login/LogincScene.ts (hunk 8 above)
Comment (gparant) [reply front/src/Phaser/Login/LogincScene.ts:161]: I'm not sure, but we have `GameSceneInterface` and it should represent all of them. For me, it's not a problem, and we could imagine a map loading or what ever can to use a function in this interface.

File: back/src/Model/Websocket/ExtRooms.ts (hunk 10 above)
Comment (gparant) [reply back/src/Model/Websocket/ExtRooms.ts:47]: roomId is already saved in the socket.  I have made this code here, because it's here that you refresh mapping position of all users. And in this code we save for each room each position of users.
Here we are in the Rooms socket what represent all sockets created and activated in the back.

Comment (moufmouf) [reply back/src/Model/Websocket/ExtRooms.ts:47]: Oops sorry, shouldn't be drinking and reviewing PR at the same time :D

File: README.md (hunk 7 above)
Comment (moufmouf) [reply README.md]: :+1:
